Packages/com.unity.inputsystem/InputSystem/Plugins/Switch/NPad.cs

```diff
@@ -343,16 +344,112 @@ public static NPadDeviceIOCTLOutputCommand Create()
+
+        [InputControl(name = "rightStick", format = "VC2S", layout = "Stick")]
+        [InputControl(name = "rightStick/x", offset = 0, format = "USHT", parameters = "normalize,normalizeMin=0.15,normalizeMax=0.85,normalizeZero=0.5")]
+        [InputControl(name = "rightStick/left", offset = 0, format = "USHT", parameters = "normalize,normalizeMin=0.15,normalizeMax=0.85,normalizeZero=0.5,clamp,clampMin=0,clampMax=0.5,invert")]
+        [InputControl(name = "rightStick/right", offset = 0, format = "USHT", parameters = "normalize,normalizeMin=0.15,normalizeMax=0.85,normalizeZero=0.5,clamp,clampMin=0.5,clampMax=1")]
+        [InputControl(name = "rightStick/y", offset = 2, format = "USHT", parameters = "invert,normalize,normalizeMin=0.15,normalizeMax=0.85,normalizeZero=0.5")]
+        [InputControl(name = "rightStick/up", offset = 2, format = "USHT", parameters = "normalize,normalizeMin=0.15,normalizeMax=0.85,normalizeZero=0.5,clamp,clampMin=0.15,clampMax=0.5,invert")]
+        [InputControl(name = "rightStick/down", offset = 2, format = "USHT", parameters = "normalize,normalizeMin=0.15,normalizeMax=0.85,normalizeZero=0.5,clamp,clampMin=0.5,clampMax=0.85,invert=false")]
+        [FieldOffset(8)] public ushort rightStickX;
+        [FieldOffset(10)] public ushort rightStickY;
+
+        public float leftTrigger => ((buttons & (1 << (int)Button.ZL)) != 0) ? 1f : 0f;
+
+        public float rightTrigger => ((buttons & (1 << (int)Button.ZR)) != 0) ? 1f : 0f;
+
+        public enum Button
+        {
+            North = 11,
+            South = 8,
+            West = 10,
+            East = 9,
+
+            StickL = 18,
+            StickR = 19,
+            L = 12,
+            R = 13,
+
+            ZL = 14,
+            ZR = 15,
+            Plus = 17,
+            Minus = 16,
+
+            X = North,
+            B = South,
+            Y = West,
+            A = East,
+        }
+
+        public NPadInputStateHID WithButton(Button button, bool value = true)
+        {
+            var bit = (uint)1 << (int)button;
+            if (value)
+                buttons |= bit;
+            else
+                buttons &= ~bit;
+            // dpad default state
+            buttons |= 8 << 24;
+            leftStickX = 0x8000;
+            leftStickY = 0x8000;
+            rightStickX = 0x8000;
+            rightStickY = 0x8000;
+            return this;
+        }
+    }
+#endif
 }
 
 namespace UnityEngine.InputSystem.Switch
 {
+#if UNITY_EDITOR || UNITY_SWITCH
     /// <summary>
     /// An NPad controller for Switch, which can be a Joy-Con.
     /// </summary>
     /// <seealso cref="NPadInputState"/>
-    [InputControlLayout(stateType = typeof(NPadInputState), displayName = "Switch Controller (on Switch)")]
-    public class NPad : Gamepad, INPadRumble
+    [InputControlLayout(stateType = typeof(NPadInputStateSwitch), displayName = "Switch Controller (on Switch)")]
+    public class NPadSwitch : Gamepad, INPadRumble
```

Comment:
I wonder about the naming here. Think it's unclear what `NPadSwitch` and `NPadHID` really are. Is it Joy-Cons? Is it Switch Pro controllers? Is it both?

From what I understand, `NPadSwitch` is actually both, i.e. might be Joy-Cons, might be a Pro controller. It's basically "controller picked up through Switch SDK API". Whereas `NPadHID` specifically is the Switch Pro controller (and thus has no support for orientations and such). Is that correct?

If so, I think it'd be good to make that distinction clearer in the API. It's different from DualShockGamepad, where the PS4 and HID version really are one of the same controller with the same capabilities (minus the fact that we are unable to surface all of them on the HID version).

Reply:
Yeah, I was unsure about the naming as well, also because NPad doesn't seem to be a name used in consumer branding (?). I guess we *could* support the Joy-Cons on Desktop as well, but not sure if there is much value in doing so. 

What do you propose? Should I just rename `NPadHID` to `SwitchProControllerHID`, and leave `NPadSwitch` as it was (`NPad`)?

Reply:
Have only done reading so far and no actual investigation but it seemed like chances for making JoyCon work on desktop would be slim. Didn't seem like the device is even a HID IIRC.

>What do you propose? Should I just rename NPadHID to SwitchProControllerHID, and leave NPadSwitch as it was (NPad)?

`SwitchProControllerHID` sounds good to me. Long but clear and unambiguous.

For `NPad`, personally, I'd prefer `SwitchController`. Wasn't too fond of `NPad`.

Sidetrack: I think the PS4 controller stuff suffers from similar problems. I think having `Gamepad` and then `XboxGamepad` and `PlayStationGamepad` and `SwitchGamepad` etc. would be much clearer. What do you think? Should we do a naming pass on those? We could have aliases to keep existing bindings working as is.

Reply:
As we just discussed, JoyCon support on Desktop should be possible. Listed it for nice to have >1.0.

Also, as discussed, renamed NPadHID to SwitchProControllerHID, and left NPad to be renamed later.

Reply:
FYI: NPad is an SDK internal name so it can be a common name to Switch developers, but not PC develoers. Anyway, SwitchController sounds good to me. Sorry for my sense for names!